# Inicio

In [ ]:
#%pip install pandas matplotlib

import pandas as pd
import matplotlib.pyplot as plt


[notice] A new release of pip is available: 26.0.1 -> 26.1.2
[notice] To update, run: C:\Users\carlo\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable
  Using cached cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)
  Using cached pyparsing-3.3.2-py3-none-any.whl.metadata (5.8 kB)
   ---------------------------------------- 0.0/9.8 MB ? eta -:--:--
   ------------ --------------------------- 3.1/9.8 MB 19.3 MB/s eta 0:00:01
   --------------------------------- ------ 8.1/9.8 MB 21.5 MB/s eta 0:00:01
   ---------------------------------------- 9.8/9.8 MB 21.7 MB/s  0:00:00
   ---------------------------------------- 0.0/8.2 MB ? eta -:--:--
   ------------------------- -------------- 5.2/8.2 MB 25.8 MB/s eta 0:00:01
   ---------------------------------------- 8.2/8.2 MB 24.4 MB/s  0:00:00
Using cached cycler-0.12.1-py3-none-any.whl (8.3 kB)
   ---------------------------------------- 0.0/2.3 MB ? eta -:--:--
   ------------------------------- -------- 1.8/2.3 MB 25.6 MB/s eta 0:00:01
   ------------------------------- -------- 1.8/2.3 MB 25.6 MB/s eta 0:00:01

In [5]:
print(pd.__version__)

3.0.3


In [22]:
df = pd.read_csv('server_logs.csv')
print(f'Shape: {df.shape}')

Shape: (5795, 15)


In [21]:
print(f'Info:\n')
df.info()

Info:

<class 'pandas.DataFrame'>
RangeIndex: 5795 entries, 0 to 5794
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   timestamp_event  5795 non-null   str  
 1   received_at      5795 non-null   str  
 2   service_name     5795 non-null   str  
 3   severity         5795 non-null   str  
 4   message          5795 non-null   str  
 5   trace_id         5795 non-null   str  
 6   request_id       5795 non-null   str  
 7   method           5795 non-null   str  
 8   endpoint         5795 non-null   str  
 9   status_code      5795 non-null   int64
 10  latency_ms       5795 non-null   int64
 11  host             5795 non-null   str  
 12  env              5795 non-null   str  
 13  region           5795 non-null   str  
 14  log_type         5795 non-null   str  
dtypes: int64(2), str(13)
memory usage: 679.2 KB


In [ ]:
print(f'Dtypes:\n{df.dtypes}')

Dtypes:



timestamp_event      str
received_at          str
service_name         str
severity             str
message              str
trace_id             str
request_id           str
method               str
endpoint             str
status_code        int64
latency_ms         int64
host                 str
env                  str
region               str
log_type             str
dtype: object

In [ ]:
print('Priemros 5 registros:\n')
df.head()

Priemros 5 registros:



,timestamp_event,received_at,service_name,severity,message,trace_id,request_id,method,endpoint,status_code,latency_ms,host,env,region,log_type
0,2026-01-10T00:02:39.029160Z,2026-01-10T00:02:39.097160Z,orders-service,INFO,Request completed,bd8e6ebe717b4c43ba5e72e1668641a8,e6b3009adcad,GET,/orders/create,200,99,orders-service-pod-03,prod,sa-east-1,request
1,2026-01-10T00:02:46.081021Z,2026-01-10T00:02:46.196021Z,api-gateway,INFO,Background job completed,40ac5ff97bae43d5b8045484725e0aca,e2aa1cccd1cf,GET,/health,200,122,api-gateway-pod-01,prod,sa-east-1,request
2,2026-01-10T00:04:01.648849Z,2026-01-10T00:04:01.718849Z,notification-service,WARN,Rate limit nearing threshold,c048b378759a4c54a6f7d7251a6acc88,c74106ca8581,POST,/notify/sms,200,646,notification-service-pod-03,prod,sa-east-1,request
3,2026-01-10T00:05:08.148346Z,2026-01-10T00:05:08.236346Z,api-gateway,INFO,Background job completed,072169f708c84227986bfda9f9657bdc,bf10fa3609bd,GET,/checkout,200,127,api-gateway-pod-03,prod,sa-east-1,request
4,2026-01-10T00:05:19.590837Z,2026-01-10T00:05:19.632837Z,inventory-service,INFO,Health check OK,e76a214131d24a4ea1acf389362e34cb,f672d669eda7,GET,/inv/release,200,144,inventory-service-pod-02,prod,sa-east-1,request


In [23]:
# TRansformamos el timestamp_event a datetime y nos aseguramos de que se transforme correctamente.
df["timestamp_event"] = pd.to_datetime(df["timestamp_event"])
df.dtypes

timestamp_event    datetime64[us, UTC]
received_at                        str
service_name                       str
severity                           str
message                            str
trace_id                           str
request_id                         str
method                             str
endpoint                           str
status_code                      int64
latency_ms                       int64
host                               str
env                                str
region                             str
log_type                           str
dtype: object

# Exploración inicial.

### ¿Cuántos logs hay en total?

In [43]:
df.shape[0]

5795

### En total hay 5795 logs

### ¿Qué severidad aparece más?

In [42]:
sever = df['severity']
sever.value_counts()

severity
INFO        3542
WARN        1358
ERROR        775
CRITICAL     120
Name: count, dtype: int64

### La severidad que más aparece es INFO.

### ¿Qué servicio genera más logs?

In [41]:
df['service_name'].value_counts()

service_name
api-gateway             1509
orders-service          1057
inventory-service        964
payment-service          842
auth-service             778
notification-service     645
Name: count, dtype: int64

### El servicio que genera más logs es api-gateway.

### ¿Qué servicio genera menos logs?

### El servicio que genera menos logs es notification-service.

### ¿Cuál es el mensaje más repetido?

In [44]:
print(df['message'].value_counts())

message
Health check OK                                   1196
Background job completed                          1185
Request completed                                 1161
Order creation failed - inventory lock timeout     197
Rate limit nearing threshold                       193
Elevated 401 responses at gateway                  193
Retrying upstream request                          184
Slow response detected                             183
Cannot send receipt - Payment pending              170
Inventory endpoint degraded - high latency         114
Payment gateway unavailable                        103
Database deadlock detected                          99
Inventory reservation retry                         98
Checkout failed - upstream payment timeout          88
Increased 5xx responses detected at gateway         82
Order created but payment pending                   71
Suspicious login rate detected                      70
Possible credential stuffing detected               69
Pa

### El mensaje más repetido es "Health check OK".

### ¿Cuál es el mensaje “malo” más repetido?

In [45]:
bad_event = df[ (df["severity"] == "ERROR") | (df["severity"] == "CRITICAL") | (df["status_code"] >= 500)]
bad_event['message'].value_counts()

message
Order creation failed - inventory lock timeout    197
Payment gateway unavailable                       103
Database deadlock detected                         99
Checkout failed - upstream payment timeout         88
Possible credential stuffing detected              69
Payment Gateway Timeout - 504 Gateway Time-out     67
Database query failed                              62
External dependency error                          59
Invalid credentials - 401 spike                    54
Upstream timeout                                   46
Out of memory                                      29
Service crash detected                             22
Name: count, dtype: int64

### El Bad Event más repetido es:
#### Order creation failed - inventory lock timeout
#### Con 197 apariciones.